docs/internal/constrained-memory.md

```diff
@@ -0,0 +1,175 @@
+
+Currently, Verona lacks an equivalent of C's `sizeof` operator and so does not have a way of specifying the amount of memory required in terms of something relevant to the programmer.
+It might be possible for the constructor for such a region to be a generic that took a type list.
+For example:
+
+```verona
+try
+{
+  var region = PrereservedRegion[Array[SomeCls, 42], SomeBigCls, Array[SomeSmallCls, 1000]];
+  allocateABunchOfStuff(region);
+}
+catch
+{
+  { OutOfMemory => /* report failure here */ }
+}
+```
+
+### Low-memory noticeboards
+
+XNU, Android, and Windows all provide some form of low-memory notifications.
+The most useful forms of these are when memory is fairly low but allocation remains possible.
+These can be used by software to avoid allocation or free existing objects when memory is low but not yet exhausted.
+These are particularly useful in garbage-collected systems where it's often beneficial to avoid running the GC until the system starts to experience some memory pressure.
+
+In a Verona world, traced regions are only collected in response to explicit invocation of the collector.
+A cown may wish to avoid tracing until memory starts to be constrained.
+Signalling a noticeboard and scheduling the cowns that have requested notifications does not require memory allocation and so providing a low-memory noticeboard would allow a cown to perform some cleanup when memory is low, without requiring any additional allocation.
+
+### Sudden-termination cowns
+
+The concept of sudden termination could be extended to `cown`s.
+Cowns can already be held by weak references and so this would require a small tweak to the weak reference semantics such that cowns supporting sudden termination are not deallocated until their weak and strong reference count reference counts *both* drop to zero, or until an out-of-memory condition is reached.
+
+When a cown is running, it can opt into sudden termination.
+This would add it to a list in the runtime.
+When the runtime detects an out-of-memory condition then it would walk the list of cowns that support sudden termination and free any that are not currently strongly referenced.
+
+Cown references are not required to be option types and so there is no way of making cowns that can go away unexpectedly other than by ensuring that all references to it are weak references.
+Scheduling a behaviour on a cown implicitly references it and so a cown that supports sudden termination will not be deallocated if it has pending behaviours to process or if it is running.
+
+When a `cown` is deallocated (either explicitly or when killed by low-memory conditions) then it will need to be removed from the list.
+This is likely to be a fairly uncommon operation and so it would probably be acceptable to have a simple doubly linked list protected by a lock.
+
+In the current implementation, `cown`s are not deallocated while they still have strong references but all objects that they point to are.
```

Comment:
That seems sketchy; I must be misunderstanding... what happens to pending behaviours or what's given to behaviours that get scheduled?  Or is it just objects weakly held that are discarded?

Reply:
Sorry, terminology problem:

Cowns in the abstract machine are dealloated as soon as the last strong reference to them goes away.

The cown data structure in the runtime is not deallocated until all weak references to it have gone away.  

This is basically the same as `std::shared_ptr` created with `std::make_shared`: the refcounting control block is part of the allocation and so it can't be deallocated until the object is deallocated.  The finaliser runs and it's gone from the abstract machine, but the memory is not reclaimed.